app/protocol/flowcontext/should_mine.go

```diff
@@ -35,13 +26,15 @@ func (f *FlowContext) ShouldMine() (bool, error) {
 	}
 
 	now := mstime.Now().UnixMilliseconds()
```

Comment:
The queries to virtual selected parent info are relatively expensive. It might make sense to maintain the `virtual selected parent timestamp` by pushing virtual changes and only querying the current time against it with out acquiring the consensus lock.